### PR TITLE
Performance boost for Raycast

### PR DESCRIPTION
**Pull request by sfence**
Performance boost for Raycast by ignoring node boxes smaller than node wherever possible.

## To do

This PR is Ready for Review.

## How to test

Try something like this (by @appgurueu ) on master and this branch:

```
minetest.register_chatcommand("rays", {
	description = "Do the rays",
	func = function(name)
		local t = minetest.get_us_time()
		local p = minetest.get_player_by_name(name)
		for i = 1, 1000 do
			for thing in minetest.raycast(p:get_pos(), p:get_pos():offset(math.random(-100, 100), math.random(-100, 100), math.random(-100, 100)), false, false) do
			end
		end
		return true, (minetest.get_us_time() - t) .. "us" 
	end
})
``` 

On the surface in devTest:
pr

### CODE REVIEW

File: src/environment.cpp [hunk 1]
```diff
@@ -191,6 +196,11 @@ void Environment::continueRaycast(RaycastState *state, PointedThing *result_p)
 			if (!is_valid_position)
 				continue;
 
+			// if take sense to use it
+			if ( (cy != y || cx != x || cz != z)
```
Comment (appgurueu) [inline src/environment.cpp:200]: How about just doing `state->m_iterator.m_current_node_pos != np`? Then we can also get rid of the three variables above.

File: doc/lua_api.md [hunk 2]
```diff
@@ -9702,6 +9702,8 @@ Used by `minetest.register_node`.
 
     selection_box = {
         -- see [Node boxes] for possibilities
+        -- Selection boxes that oversize node size can cause
+        -- signifiant performance drop of Raycasts.
```
Comment (appgurueu) [inline doc/lua_api.md:9706]:
```suggestion
        -- significant performance drop of Raycasts.
```

File: src/nodedef.cpp [hunk 3]
```diff
@@ -1311,6 +1311,16 @@ inline void NodeDefManager::fixSelectionBoxIntUnion()
 		m_selection_box_union.MaxEdge.Z / BS - 0.5f);
 }
 
+inline void NodeDefManager::calcBigSelectinBox(content_t id, const ContentFeatures &def)
+{
+	aabb3f box_union;
+	getNodeBoxUnion(def.selection_box, def, &box_union);
```
Comment (appgurueu) [inline src/nodedef.cpp:1317]: Can't we reuse `m_selection_box_union`?

File: src/nodedef.cpp [hunk 4]
```diff
@@ -1311,6 +1311,16 @@ inline void NodeDefManager::fixSelectionBoxIntUnion()
 		m_selection_box_union.MaxEdge.Z / BS - 0.5f);
 }
 
+inline void NodeDefManager::calcBigSelectinBox(content_t id, const ContentFeatures &def)
```
Comment (appgurueu) [inline src/nodedef.cpp:1314]:
```suggestion
inline void NodeDefManager::calcBigSelectionBox(content_t id, const ContentFeatures &def)
```

File: src/nodedef.cpp [hunk 5]
```diff
@@ -1360,12 +1370,14 @@ content_t NodeDefManager::set(const std::string &name, const ContentFeatures &de
 	m_content_features[id] = def;
 	m_content_features[id].floats = itemgroup_get(def.groups, "float") != 0;
 	m_content_lighting_flag_cache[id] = def.getLightingFlags();
+
+	calcBigSelectinBox(id, def);
```
Comment (appgurueu) [inline src/nodedef.cpp:1374]: I would put this next to `fixSelectionBoxIntUnion`

File: src/environment.cpp (hunk 1 above)
Comment (SmallJoker) [reply src/environment.cpp:200]: Or `v3s16 cp = state->m_iterator.m_current_node_pos;` if the goal was to move the to the stack for performance reasons.

File: src/nodedef.cpp [hunk 6]
```diff
@@ -1575,6 +1587,9 @@ void NodeDefManager::deSerialize(std::istream &is, u16 protocol_version)
 		m_content_features[i].floats = itemgroup_get(f.groups, "float") != 0;
 		m_content_lighting_flag_cache[i] = f.getLightingFlags();
 		addNameIdMapping(i, f.name);
+
+		calcBigSelectinBox(i, f);
```
Comment (appgurueu) [inline src/nodedef.cpp:1591]: Same here

File: src/nodedef.h [hunk 7]
```diff
@@ -790,6 +792,12 @@ class NodeDefManager {
 	 */
 	void fixSelectionBoxIntUnion();
 
+	/*!
+	 * Calculate ContentFeatures bigSelectonBox value.
+	 * m_selection_box_union.
```
Comment (appgurueu) [inline src/nodedef.h:797]: Please touch this comment up a bit. I think just `/// Calculate ContentFeatures bigSelectionBox value.` would suffice.

File: src/nodedef.cpp (hunk 3 above)
Comment (sfence) [reply src/nodedef.cpp:1317]: No. `m_selection_box_union` is a union over all already registered nodes.
This should be per node bool.

Comment (appgurueu) [reply src/nodedef.cpp:1317]: sorry brainfart

**Review by appgurueu — APPROVED** [review]
There are still a couple typos but otherwise this is fine, works and yields the speedups you claim.

File: src/nodedef.h [hunk 8]
```diff
@@ -441,6 +441,8 @@ struct ContentFeatures
 	NodeBox node_box;
 	NodeBox selection_box;
 	NodeBox collision_box;
+	// if selection box is bigger then (-BS,-BS,-BS)/2 and (BS,BS,BS)/2 edges
+	bool bigSelectionBox;
```
Comment (SmallJoker) [inline src/nodedef.h:445]:
```suggestion
	/// if selection box is bigger then (-BS,-BS,-BS)/2 and (BS,BS,BS)/2 edges
	bool have_big_selection_box;
```
1. https://dev.minetest.net/Code_style_guidelines#Miscellaneous
   (The "local" part of "local variable" could be removed eventually)
2. Doxygen formatting for better IDE experience

File: src/nodedef.cpp [hunk 9]
```diff
@@ -1311,6 +1311,16 @@ inline void NodeDefManager::fixSelectionBoxIntUnion()
 		m_selection_box_union.MaxEdge.Z / BS - 0.5f);
 }
 
+inline void NodeDefManager::calcBigSelectionBox(content_t id, const ContentFeatures &def)
+{
+	aabb3f box_union;
+	getNodeBoxUnion(def.selection_box, def, &box_union);
+	bool bigSelectionBox =
```
Comment (SmallJoker) [inline src/nodedef.cpp:1318]: Naming here too.

File: src/mapnode.h [hunk 10]
```diff
@@ -293,6 +293,11 @@ struct alignas(u32) MapNode
 	void getCollisionBoxes(const NodeDefManager *nodemgr,
 		std::vector<aabb3f> *boxes, u8 neighbors = 0) const;
 
+	/*
+		Return true if Have big selection boxes
+	*/
+	bool haveBigSelectionBox(const NodeDefManager *nodemgr) const;
```
Comment (SmallJoker) [inline src/mapnode.h:299]:
```suggestion
	/// Returns `true` when an extent is > BS/2.
	bool haveBigSelectionBox(const NodeDefManager *nodemgr) const;
```

Also: this function might be superfluous. `Environment` already has access to NodeDefManager and MapNode to retrieve this value when needed.

File: src/environment.cpp [hunk 11]
```diff
@@ -179,6 +179,9 @@ void Environment::continueRaycast(RaycastState *state, PointedThing *result_p)
 			break; // About to go out of bounds
 		}
 
+		// for check bigSelectionBox
+		v3s16 &current_pos = state->m_iterator.m_current_node_pos;
```
Comment (SmallJoker) [inline src/environment.cpp:183]:
```suggestion
		// Optimization trick: to skip non-oversized selection boxes for other positions.
		const v3s16 current_pos = state->m_iterator.m_current_node_pos;
```

**Review by SmallJoker — COMMENTED** [review]
Aesthetics only.

**Comment by Zughy** [comment]
@sfence please reopen if you want to address SmallJoker's review (so sad for me to close a PR like this, as I use many raycasts)

File: src/nodedef.h (hunk 8 above)
Comment (appgurueu) [reply src/nodedef.h:445]: fixed

File: src/nodedef.cpp (hunk 9 above)
Comment (appgurueu) [reply src/nodedef.cpp:1318]: fixed

File: src/mapnode.h (hunk 10 above)
Comment (appgurueu) [reply src/mapnode.h:299]: > Also: this function might be superfluous.

It is. Good call, removed it.

File: src/environment.cpp (hunk 11 above)
Comment (appgurueu) [reply src/environment.cpp:183]: Done. Our conventions say we want to copy such small vectors anyways (and we do want to).

Side note: I would consider it likely that both end up with the same machine code (at least if a const ref was used), that is, the compiler would figure this out itself.

**Comment by appgurueu** [comment]
Note: when squashing this, please remove me from the co-authors, for it is really sfence's code in all important aspects

**Review by SmallJoker — APPROVED** [review]
Thank you @appgurueu for taking care of the open points.

Performance testing results, with maximal FPS for more samples:

. | Time master / µs | Time PR / µs
-- | -- | --
Average | 18.00 | 12.33
Deviation | 2.43 | 2.14
Median time | 18.00 | 12.50

18-19 measurement samples from the F6 debugger using the following code:

```diff
diff --git a/src/client/game.cpp b/src/client/game.cpp
index dc13095d1..df28a4499 100644
--- a/src/client/game.cpp
+++ b/src/client/game.cpp
@@ -3491,7 +3491,12 @@ PointedThing Game::updatePointedThing(
 
        RaycastState s(shootline, look_for_object, liquids_pointable, pointabilities);
        PointedThing result;
-       env.continueRaycast(&s, &result);
+       {
+               ScopeProfiler sp(g_profiler, __func__, SPT_AVG);
+               env.continueRaycast(&s, &result);
+               if ((rand() & 0xfff) == 0)
+                       printf("Distance: %.3f\n", sqrtf(result.distanceSq) / BS);
+       }
        if (result.type == POINTEDTHING_OBJECT) {
                hud->pointing_at_object = true;
 
```
Mod:
```lua
core.register_craftitem("lab:long_long_man", {
	description = "Long poiny thing",
	inventory_image = "object_marker_red.png",
	range = 100
})
```
Pointing range: `Distance: 96.776` (same for both measurements)

**Comment by SmallJoker** [comment]
Actually I have to circle back on my judgement. Now it seems to skip too much:

grafik
grafik

**Pull request merged**